libfwupdplugin/fu-device.c

```diff
@@ -2472,30 +2473,114 @@ fu_device_add_flag (FuDevice *self, FwupdDeviceFlags flag)
 		fu_device_inhibit (self, "needs-activation", "Pending activation");
 }
 
+typedef struct {
+	guint64		*bitfield;
+	guint64		 value;
+	gchar		*value_str;
+} FuDeviceCustomFlagItem;
+
+static void
+fu_device_custom_flag_free (FuDeviceCustomFlagItem *item)
+{
+	g_free (item->value_str);
+	g_free (item);
+}
+
+static FuDeviceCustomFlagItem *
+fu_device_custom_flag_find_by_str (FuDevice *self, const gchar *value_str)
+{
+	FuDevicePrivate *priv = GET_PRIVATE (self);
+	if (priv->custom_flags == NULL)
+		return NULL;
+	for (guint i = 0; i < priv->custom_flags->len; i++) {
+		FuDeviceCustomFlagItem *item = g_ptr_array_index (priv->custom_flags, i);
+		if (g_strcmp0 (item->value_str, value_str) == 0)
+			return item;
+	}
+	return NULL;
+}
+
+/**
+ * fu_device_register_custom_flag:
+ * @self: a #FuDevice
+ * @bitfield: (out): bitfield to control
+ * @value: an integer value
+ * @value_str: a string that represents @value
+ *
+ * Registers a custom device flag that can be used from quirk files.
+ *
+ * Since: 1.6.2
+ **/
+void
+fu_device_register_custom_flag (FuDevice *self,
+				guint64 *bitfield,
```

Comment:
Hmm, I think it's a little bit weird to send in a bitfield pointer that gets written to.  Why not instead just have a guint64 bitfield in the local device that you assign meaning to each bit when you make the custom flag?  It's very similar to what you're doing already but the ownership would be with the `fu-device`.

Reply:
Then we'd have `fu_device_add_flag()`, `fu_device_add_internal_flag()` and presumably something like `fu_device_add_private_flag` which gets a bit confusing, no? We'd also need `fu_device_remove_private_flag()`, `fu_device_get_private_flags()` etc too. My idea of using the bitfield pointer is that we can use the thing _already_ being used by the plugin.

Reply:
But what happens if you need to refer to the private flags _between_ two plugins?  Maybe I answered my own Q...  that makes the stronger argument for an internal flag in that case.

Reply:
> But what happens if you need to refer to the private flags between two plugins

I don't think we want to allow that.

Reply:
Okay this approach looks good to me then. I think the doc on the register function needs a little more meat to explain how it works though.

Reply:
I'll port some more plugins and see if the approach works there. Thanks for the sanity check.

Reply:
>  Why not instead just have a guint64 bitfield in the local device

I think you might be right.

Reply:
Narrator: _he was right_.